src/Model/Host.php

```diff
@@ -21,59 +21,91 @@
 
 namespace Friendica\Model;
 
-use Friendica\Core\Logger;
-use Friendica\Database\DBA;
+use Friendica\Database\Database;
+use Psr\Log\LoggerInterface;
 
 class Host
 {
+	/** @var string db table of host model */
+	const TABLE = 'host';
+
 	/**
-	 * Get the id for a given hostname
-	 * When empty, the current hostname is used
+	 * Defines the environment variable, which includes the current node name instead of the detected hostname
 	 *
-	 * @param string $hostname
+	 * @var string
 	 *
-	 * @return integer host name id
-	 * @throws \Exception
+	 * @notice This is used for cluster environments, where the each node defines it own hostname.
 	 */
-	public static function getId(string $hostname = '')
+	const ENV_VARIABLE = 'NODE_NAME';
+
+	/** @var integer The host id */
+	private $id = -1;
+	/** @var string The host name */
+	private $name;
+
+	/** @var LoggerInterface */
+	private $logger;
+
+	public function __construct(Database $dba,  LoggerInterface $logger, array $server = [])
+	{
+		$this->logger = $logger;
+
+		$this->detectHostname($dba, $server);
+	}
+
+	private function detectHostname(Database $dba, array $server)
 	{
+		$hostname = $server[self::ENV_VARIABLE] ?? null;
+
 		if (empty($hostname)) {
 			$hostname = php_uname('n');
 		}
 
-		$hostname = strtolower($hostname);
-
-		$host = DBA::selectFirst('host', ['id'], ['name' => $hostname]);
-		if (!empty($host['id'])) {
-			return $host['id'];
+		// Trim whitespaces first to avoid getting an empty hostname
+		// For linux the hostname is read from file /proc/sys/kernel/hostname directly
+		$hostname = trim($hostname);
+		if (empty($hostname)) {
+			$this->logger->error('Empty hostname is invalid.');
+			$this->id = -1;
+			$this->name = "";
+			return;
 		}
 
-		DBA::replace('host', ['name' => $hostname]);
+		$this->name = strtolower($hostname);
 
-		$host = DBA::selectFirst('host', ['id'], ['name' => $hostname]);
-		if (empty($host['id'])) {
-			Logger::warning('Host name could not be inserted', ['name' => $hostname]);
-			return 0;
-		}
+		$host = $dba->selectFirst(self::TABLE, ['id'], ['name' => $this->name]);
+		if (!empty($host['id'])) {
+			$this->id = (int)$host['id'];
+		} else {
+			$dba->replace(self::TABLE, ['name' => $hostname]);
 
-		return $host['id'];
+			$host = $dba->selectFirst(self::TABLE, ['id'], ['name' => $hostname]);
+			if (empty($host['id'])) {
+				$this->id = -1;
+				$this->logger->warning('Host name could not be inserted', ['name' => $hostname]);
+			} else {
+				$this->id = (int)$host['id'];
+			}
+		}
 	}
 
 	/**
-	 * Get the hostname for a given id
+	 * Get the id for a given host
 	 *
-	 * @param int $id
+	 * @return integer host name id
+	 */
+	public function getId()
```

Comment:
Could you extend this function so that it can return the id for a given hostname as well?

Reply:
What would be the use?

Reply:
When there are several machines that are running the worker queues, they could be different in their performance. So you might want to limit the number of parallel worker tasks different per machine. So you would need a configuration for that by hostname.

Reply:
Then we'd probably need a separate Repository class to retrieve a `Host` model by name/id, and detect and instantiate the local `Host` object. The `Model\Host` class would probably become an Active Record like `Model\Introduction`.

Reply:
A model class is responsible for representing a specific host of the persistence layer. So for me the `getId()` is responsible for the concrete Id of this concrete host row. I don't want to add any way to override the inner state of this object. Thins like decide the amount of workers based on hosts is a responsibility of the method which is **using** the Host, but not the host itself.

So I'd prefer having two different ways per call:
* A unique `Host` class, which will get called everytime "host" is retrieved during a call
  * per `DI::host()`
  * if automatically called per Dice inside a constructor (like for the constructor in the `LockFactory`)
* A `Repository\Host` class for selecting or altering host entries
  * per `DI::hostRepo()`

Reply:
Is `Repository` the new `Model`? Because until now I thought that `Model` is - for example - the interface between code and database for a specific table?

Reply:
The `Repository` is the interface between a **whole table** and the code
The `Model` is the interface between **one or many rows** of a table and the code